src/coord/coord.rs

```diff
@@ -1650,8 +1656,73 @@ where
+        self.validate_dataflow(&mut dataflow);
+        self.broadcast_dataflow_creation(dataflow);
+    }
+
+    /// Performs some sanity checking on the dataflow before shipping it.
+    ///
+    /// In particular, there are requirement on the `as_of` field for the dataflow
+    /// and the `since` frontiers of created arrangements, as a function of the `since`
+    /// frontiers of dataflow inputs (sources and imported arrangements).
+    fn validate_dataflow(&mut self, dataflow: &mut DataflowDesc) {
+        use differential_dataflow::lattice::Lattice;
+        use timely::progress::timestamp::Timestamp;
+
+        // The identity for `join` is the minimum element.
+        let mut since = Antichain::from_elem(Timestamp::minimum());
+
+        // TODO: Populate "valid from" information for each source.
+        // For each source, ... do nothing because we don't track `since` for sources.
+        // for (instance_id, _description) in dataflow.source_imports.iter() {
+        //     // TODO: Extract `since` information about each source and apply here. E.g.
+        //     since.join_assign(&self.source_info[instance_id].since);
+        // }
+
+        // For each imported arrangement, lower bound `since` by its own frontier.
+        for (global_id, (_description, _typ)) in dataflow.index_imports.iter() {
+            since.join_assign(
+                self.indexes
+                    .since_of(global_id)
+                    .expect("global id missing at coordinator"),
+            );
+        }
+
+        // For each produced arrangement, force its compaction frontier to be at least `since`.
+        for (global_id, _description, _typ) in dataflow.index_exports.iter() {
+            self.indexes
+                .get_mut(global_id)
+                .expect("global id missing at coordinator")
+                .advance_since(&since);
+        }
+
+        // TODO: Produce "valid from" information for each sink.
+        // For each sink, ... do nothing because we don't yield `since` for sinks.
+        // for (global_id, _description) in dataflow.sink_exports.iter() {
+        //     // TODO: assign `since` to a "valid from" element of the sink. E.g.
+        //     self.sink_info[global_id].valid_from(&since);
+        // }
+
+        // Ensure that the dataflow's `as_of` is at least `since`.
+        if let Some(as_of) = &mut dataflow.as_of {
+            // If we have requested a specific time that is invalid .. someone errored.
+            use timely::order::PartialOrder;
+            if !(<_ as PartialOrder>::less_equal(&since, as_of)) {
+                // This can occur in SINK and TAIL at the moment. Their behaviors are fluid enough
+                // that we just correct to avoid producing incorrect output updates, but we should
+                // fix the root of the problem in a more principled manner.
+                log::error!(
+                    "Requested as_of ({:?}) not >= since ({:?}); correcting",
```

Comment:
Lets also print the dataflow desc's `debug_name` here?

Reply:
Man, we need a !@#$ CI merge plz button.